North-Seattle-College/ad440-winter2022-thursday-repo
# Enrich conversation data with sentiment/question/emotion metadata

Closes Issue #91 .

## Description
- Creates script to analyze each conversation message and adds metadata for if the message text is a question/sentence, sentiment analysis (text is positive/negative/neutral), and emotion analysis (text is happy, sad, etc).
- Script accesses the floop conversation data json stored in S3, analyzes the conversation messages contained within, and uploads a new json file with the metadata attached.

Additional files:
 - `emotions.txt` - updated from Payam's version with 3 additional entries.

#### Metadata Analysis Output
**Question/Sentence Analysis** - Based on Payam's `NLP.py`, will analyze the conversation text based on ending punctuation and starting question words. Output in the metadata will be one of the following:
- `question`: text either ended with a question mark or started with an indicated question word.
- `sentence`: text ended with a period or exclamation point.
- `statement`: text ended with a letter or number and no punctuation of any kind.
- `Unknown`: text ended with an unusual character or emoji.

**Sentiment Analysis** - Based on Steven's `Classification.py`, will analyze the positive/negative spectrum of the conversation text. Output of the metadata will be one of the following:
- `positive`: text was classified as positive in tone.
- `negative`: text was classified as negative in tone.
- `neutral`: text was classified as neither positive or neutral in tone.

**Emotion Analysis** - Based on Payam's `NLP.py`, will analyze the conversation text based on appearance of words and their associated emotions, as per the `emotions.txt` file. Output of the metadata will follow the below format:
- `{emotion}`: one of 541 words was detected in conversation text, and the associated emotion was returned.
- `hard to read`: no emotional tagging could b performed on the conversation text.

## Prerequisites
- **Imports:** In terminal, run the following if not already installed:
_NOTE: you may need to go to the project root folder and launch a virtual environment for the imports to be found._ 
   - `pip install nltk`
   - `pip install boto3`
- **AWS Credentials**
  - Boto3 has ([established processes for finding existing credentials](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials)). If you have not configured yours using the AWS CLI, you can do so following the instructions ([here](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)).
    - after installing the AWS CLI, you run `aws configure` and follow the prompts to enter your AWS key/secret key information.
  - You can confirm you have successfully configured your credentials by locating the credentials file at `~/.aws/credentials`, where `~` is your root folder(for example, "C:\Users\Profile").
     
## Steps to Verify
(from the backend/scripts folder)
1.  Run script in the command line by entering: `floop-conv-enrich.py`
    This will run the script, which will access the S3 bucket and `auto-floop-s3-export3.json` file. There it will load the data, and take a subset of 50 elements of the data (to speed up verification time).
2. When your command prompt returns, the script has finished.
3. Log into the AWS Console and the S3 bucket: [ad440-mpg-floop-export-storage](https://s3.console.aws.amazon.com/s3/buckets/ad440-mpg-floop-export-storage?region=us-west-2&tab=objects)
4. Verify the file `enriched-conv-sample.json` has been uploaded, noting the timestamp (should align with the time you ran the script):
![image](https://user-images.githubusercontent.com/49133101/155812388-2acebba7-5624-4650-a1a5-f28fabee8f91.png)
5. Click on the link for `enriched-conv-sample.json`.
6. in the menu of buttons underneath the file title, click the `Open` button:
![image](https://user-images.githubusercontent.com/49133101/155812587-f8cc8855-cb94-4173-9929-e19077682d5e.png)
7. In the new tab, verify the json data now has each conversation thread element with the following fields:
- `Text`
- `uid`
- `metadata`
8. Verify each `Text` has a `metadata` tag now added:
![image](https://user-images.githubusercontent.com/49133101/155814849-b4e0ed8c-7d5d-4a2b-9558-d884b232335a.png)
